### PR TITLE
[New Cog] Polling Commands

### LINKED ISSUE #27: Polling/Quiz Polls
Create a command that allows instructors to poll students. This could be used as quizzes or just questionnaires for the students.

https://github.com/SE21-Team2/ClassMateBot/issues/38

### LINKED ISSUE #108: Add test cases for polling.py
Add test cases for polling.py file

### LINKED ISSUE #109: poll command allows empty brackets
See PR #107 or branch: [group25-pooling](https://github.com/CSC510-Group-25/ClassMateBot/tree/group25-pooling)

It could probably use additional input checks.

pollbug

### LINKED ISSUE #113: polling.py: incorrectly using title brackets results in incorrect title
Obviously issue name is obvious! See screenshot below.

See PR #107 or branch: [group25-pooling](https://github.com/CSC510-Group-25/ClassMateBot/tree/group25-pooling)

polling.py

Screenshot!

pollbug3

(wow, bot... y'don't say)

### LINKED ISSUE #114: polling.py: incorrect use of option bracket results in deformed poll
See PR #107 or branch: [group25-pooling](https://github.com/CSC510-Group-25/ClassMateBot/tree/group25-pooling)

`$poll {title} [bad input [input]` 

correctly outputs an error message, whereas

`$poll {title} bad input] [input]`

results in this:

pollbug4

### LINKED ISSUE #115: polling.py: mismatched brackets not detected
See PR #107 or branch: [group25-pooling](https://github.com/CSC510-Group-25/ClassMateBot/tree/group25-pooling)

pollbug5

with input:

pollbug5b

### PR DESCRIPTION
**Pull request by peeyush10234**
### Summary

```
* Adding poll Command
* Added quizpoll Command
* Added polling tests

$poll command creates a simple reaction poll with thumbs up, thumbs down, or unsure.
$quizpoll command creates a multi-reaction poll with options A-F.

poll command tests:

    * Test poll: no input
    * Test poll: whitespace
    * Test poll: question too short
    * Test poll: student
    * Test poll: instructor
    * Test poll: reactions

quizpoll command tests:

    * Test quizpoll: no input
    * Test quizpoll: title is whitespace
    * Test quizpoll: title is too short
    * Test quizpoll: too few options
    * Test quizpoll: too many options
    * Test quizpoll: option is empty
    * Test quizpoll: embed

Co-authored-by: peeyush10234 <ptaneja@ncsu.edu>
Co-authored-by: Leila Moran <lmmoran@ncsu.edu>
```

### New Files

cogs/polling.py

$poll QUESTION
$quizpoll "TITLE" [option 1] [option2] ... [option 6] 

$poll command creates a simple reaction poll with thumbs up, thumbs down, or unsure.

$quizpoll command creates a multi-reaction poll with options A-F.

### Changed Files

test_bot.py:

* Added 13 tests for polling.py
* Fixed bare-except in test_bot.py

### Checklist

- [x] Did you test your changes locally?  
- [x] Did your changes pass all existing tests?  
- [x] Do test cases exist for your changes? If not, please make an issue.

Note: ensure that all new code has test coverage! 

----

#### Closing Issues

Resolves #27
Resolves #108 
Resolves #115
Resolves #114
Resolves #113
Resolves #109

#### Contributors

@peeyush10234 
@snapcat

### CODE REVIEW

File: cogs/polling.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                        pollMessage = await message.channel.send(embed=e)
+                        i = 0
+                        final_options = []  # There is a better way to do this for sure, but it also works that way
+                        for choice in options:
+                            if not i == len(options) and not options[i] == "":
+                                final_options.append(choice)
+                                await pollMessage.add_reaction(self.emojiLetters[i])
+                            i += 1
+                    except KeyError:
+                        return "Please make sure you are using the format 'poll: {title} [Option1] [Option2] [Option 3]'"
```
Comment (snapcat) [inline cogs/polling.py:113]: Should be:

```suggestion
                        await ctx.author.send(
                            'To use the poll command, do: $poll {POLL TITLE} [option1] [option2] [option3] ... with up to (5-8) options. \n '
                            'OR, to create an opinion poll with thumbs up/down options: $poll MESSAGE\n'
                            'EX: $poll Is this a good idea?')
                        await ctx.message.delete()
                        return
```

File: cogs/polling.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import discord
+import asyncio
```
Comment (snapcat) [inline cogs/polling.py:2]: 'asyncio' imported but unused (F401)

To shut the bot up:

```suggestion
#import asyncio
```

File: cogs/polling.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+from discord.ext import commands
+import random
```
Comment (snapcat) [inline cogs/polling.py:4]: 'random' imported but unused (F401)

To shut the bot up:

```suggestion
#import random
```

File: cogs/polling.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+from discord.ext.commands.cooldowns import BucketType
+
+
+class Poll(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+
+        self.emojiLetters = [
```
Comment (snapcat) [inline cogs/polling.py:12]: Do we need the full alphabet? I think we should limit it to A-F. We need to consider message length as well.

**Review by snapcat — CHANGES_REQUESTED** [review]
poll should be split into two commands: $poll and $quizpoll

Some bugs also need to be addressed. See: #111 and #109

File: cogs/polling.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
+import discord
+import asyncio
+from discord.ext import commands
+import random
+from discord.ext.commands.cooldowns import BucketType
+
+
```
Comment (snapcat) [inline cogs/polling.py:7]: Missing doc comment!

```suggestion
    # -----------------------------------------------------------------------------------------------------------------
    #    Function: poll(self, ctx)
    #    Description: TODO
    #    Inputs:
    #       - ctx: context of the command
    #       -
    #    Outputs:
    #       - TODO
    # -----------------------------------------------------------------------------------------------------------------
```

File: cogs/polling.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    def find_options(self, message, options):
+        # first index of the first character of the option
+        first = message.find('[') + 1
+        # index of the last character of the title
+        last = message.find(']')
+        if (first == 0 or last == -1):
+            if len(options) < 2:
+                return "Not using the command correctly"
+            else:
+                return options
```
Comment (snapcat) [inline cogs/polling.py:61]: Pylint and codacy: cogs/polling.py:58:12: R1705: Unnecessary "else" after "return" (no-else-return)

```suggestion
            return options
```

File: cogs/polling.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                            i += 1
+
+                        ads = [""]
+
+                        e = discord.Embed(title="**" + title + "**",
+                                          description=pollMessage + ads[0],
+                                          colour=0x83bae3)
+                        pollMessage = await message.channel.send(embed=e)
+                        i = 0
+                        final_options = []  # There is a better way to do this for sure, but it also works that way
```
Comment (snapcat) [inline cogs/polling.py:106]: What is final_options for? It doesn't seem to do anything.

**Review by snapcat — DISMISSED** [review]
Thanks for approving the refactor! 

I think this is now ready to go!

**Review by snapcat — APPROVED** [review]
Reverted an update from main to keep things clean. Previous approval stands!

**Review by shraddhamishra7 — APPROVED** [review]

**Pull request merged**